File: Java/src/main/java/com/DSA/Java/Graphs/SSSPP/SSSPPUsingMatrix.java
package com.DSA.Java.Graphs.SSSPP;
import java.util.LinkedList;

import com.DSA.Java.Graphs.Traversal.BFSAndDFS.AdjacencyMatrix.Graph;
import com.DSA.Java.Graphs.Traversal.BFSAndDFS.AdjacencyMatrix.GraphNode;

public class SSSPPUsingMatrix extends Graph {
    private void printPath(GraphNode node) {
        if (node.parent != null) {
            printPath(node.parent);
        }
        System.out.print(node.name + " ");
    }

    private void bfsForSSSPPHelper(GraphNode root) {
        LinkedList<GraphNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            GraphNode currentNode = queue.remove(0);
            currentNode.isVisited = true;
            System.out.print("Printing the path for " + currentNode.name + ": ");
            printPath(currentNode);
            System.out.println();

            for (GraphNode node : getNeighbours(currentNode)) {
                if (!node.isVisited) {
                    queue.add(node);
                    node.isVisited = true;
                    node.parent = currentNode;
                }
            }
        }
    }

    public void bfsForSSSPP() {
        bfsForSSSPPHelper(getNodeList().get(0));
    }
}

File: C++/Graph/src/SSSPP/SSSPPUsingMatrix.cpp
#include "../../include/SSSPP/SSSPPUsingMatrix.h"
#include <queue>

void SSSPPUsingMatrix::print_path(GraphNode *node) {
  if (node->parent != nullptr) {
    print_path(node->parent);
  }
  std::cout << node->name << " ";
}

void SSSPPUsingMatrix::bfs_for_SSSPP_helper(GraphNode *root) {
  std::queue<GraphNode *> queue;
  if (root == nullptr) return ;
  queue.push(root);

  while (!queue.empty()) {
    GraphNode *current_node = queue.front();
    current_node->is_visited = true;
    queue.pop();

    std::cout << "Printing path for " << current_node->name << ": ";
    print_path(current_node);
    std::cout << "\n";

    std::vector<GraphNode *> neighbours = get_neighbours(current_node);
    for (const auto &neighbour : neighbours) {
      if (!neighbour->is_visited) {
        queue.push(neighbour);
        neighbour->is_visited = true;
        neighbour->parent = current_node;
      }
    }
  }
}

void SSSPPUsingMatrix::bfs_for_SSSPP() {
  std::cout << "\n";
  bfs_for_SSSPP_helper(get_node_list()[0]);
}
